### PR TITLE
Automatic cloud detection and elements retrieval for AWS

### PR DESCRIPTION
**Pull request by erichs**
EC2 instances provide a unique id beginning with the string "ec2" at
/sys/hypervisor/uuid. The presence of this file, and the first three
bytes equalling "ec2" is a pretty fast and reliable method of AWS
detection at runtime.

If we're an EC2 instance, assume the http://169.254.169.254/latest
endpoint is available (also a pretty reliable bet). Crawl this
recursively for meta and user data, populating the "ec2" top-level
element path.

### CODE REVIEW

**Comment by erichs** [comment]
I admit that this branch of work is a bit more speculative, and I'd
understand if you chose not to merge it. This feature is unabashedly
provincial: it does not attempt (yet?) to support other cloud
metadata HTTP endpoints, not even ones like DigitalOcean that use
the same APIPA address (169.254.169.254)!

It has other things going against it, too:
- there's already a general-purpose mechanism for external sources.
- there's currently no way to disable it.
- it pulls in code from another project with a similar license that
  requires attribution.

So then, why?

Primarily because I felt that cloud metadata should be top-level
elements returned by the tool. Cloud metadata is important, static
data that is very likely to be useful for scripting/config mgmt.
Grafting that metadata via `elements.d` requires additional ceremony
and tooling to set up, and results in paths like:

> `external.ec2.meta_data.hostname`

and shell vars like:

> `elements_external_ec2_meta_data_hostname`

Although `hostname` is already present in the `system` path, the
result returned is not the same as the value provided by the EC2
metadata, and the EC2 version is probably the one I'm wont to use.

Addressing this important bit of configuration data via a
circuitous `external` path felt wrong to me.

Secondarily, my reasons for this feature were that it scratched an
itch, and is performant.

Lastly, I'll mention that this has the side-effect of making the
elements binary something of an EC2-detector: `elements get -p ec2`
will return "Not an EC2 instance!" under non-AWS conditions.

**Comment by jtopjian** [comment]
Very cool. In general, I think this is a great idea. All of my work is done in OpenStack, which also uses the 169 metadata URL. So EC2, OpenStack, DO - I think there are enough deployments to make this a core feature.

A few thoughts:

1. Since this is applicable to a lot more than just ec2 instances, maybe the name should be more general such as `metadata` or `cloud`. This goes for all references to "ec2".

I think `IsEc2Instance()` works fine in `utils.go`. Future detection of OpenStack et al can also go under there. Maybe making a dedicated package for that kind of detection makes sense, but I'm not worried about that right now.

hmm... I wonder if there's already a Go package that does all of this?

If not, `utils.go` works best, but ultimately all of the calls to `IsEc2Instance` will actually be `IsCloudInstance` (or something like that), which will then call `IsEc2Instance`, `IsOpenStackInstance`, etc. Does that make sense?

2.  > there's currently no way to disable it.

I think if the detection of the current system being a cloud virtual machine is accurate enough, enabling by default will be OK. I think a flag to disable querying the URL should be available, though. This does not have to be done in the initial PR.

3. > it pulls in code from another project with a similar license that requires attribution.

I am not a lawyer, but since Elements is under an MIT license, I don't think we need to embed the license into the `ec2.go` file. I think attribution to the original source should be kept - that's just good manners :)

**Comment by erichs** [comment]
> Since this is applicable to a lot more than just ec2 instances, maybe the name should be more general such as metadata or cloud. This goes for all references to "ec2".

Agree. The `cloud` top-level path gets my vote, since metadata providers often include a metadata subpath in their API. This eliminates the headaches of sorting out `metadata.meta_data`-like paths, and you just have `cloud.whatnot`.

Perhaps `IsCloudInstance()` (which calls cloud-specific test utility functions), and `GetCloudElements()`? I'll note that implementing DigitalOcean's retrieval is a gimme:

```
curl http://169.254.169.254/metadata/v1.json
```

How kind of them!

>  I think a flag to disable querying the URL should be available, though. This does not have to be done in the initial PR.

10-4!

Regarding the embedded license, I was interpreting the [parent LICENSE](https://github.com/thbishop/ec2_metadata_dump/blob/master/LICENSE) to mean that specific verbiage was required for attribution.

**Comment by erichs** [comment]
I'm less familiar with OpenStack. Is it just a matter of essentially:

```
curl http://169.254.169.254/openstack/latest/meta_data.json
```
?

**Comment by jtopjian** [comment]
> Perhaps IsCloudInstance() (which calls cloud-specific test utility functions), and GetCloudElements()? 

That sounds good to me :)

> Regarding the embedded license, I was interpreting the parent LICENSE to mean that specific verbiage was required for attribution.

My interpretation (and again, IANAL), is that since both licenses are MIT and since MIT says:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

Elements inherently already has that covered. If Elements was not MIT, I could see adding an MIT license file to cover that requirement. If you prefer having the license in the file, I'm totally fine with that.

> I'm less familiar with OpenStack. Is it just a matter of essentially:

It's the exact same as EC2:

```shell
$ curl http://169.254.169.254/latest/meta-data
ami-id
ami-launch-index
ami-manifest-path
block-device-mapping/
hostname
instance-action
instance-id
instance-type
kernel-id
local-hostname
local-ipv4
placement/
public-hostname
public-ipv4
public-keys/
ramdisk-id
reservation-id
security-groups
```

**Comment by erichs** [comment]
Ah, good catch about the LICENSE. I diff'ed them, but the differences aside from copyright were whitespace. Yeah, copyright attribution will suffice.

Re: OpenStack metadata service: does it also support `meta_data.json`?

**Comment by erichs** [comment]
Also, if you happen to have an OpenStack instance handy, what does `/sys/hypervisor/uuid`, if present, contain? What might be the best way to determine we're on an OpenStack box?

**Comment by jtopjian** [comment]
> Re: OpenStack metadata service: does it also support meta_data.json?

Doesn't look like it.

> What might be the best way to determine we're on an OpenStack box?

My understanding is the following:

```shell
$ dmidecode -s system-manufacturer
OpenStack Foundation
```

Feel free to omit OpenStack support for now -- I can add it afterwards :)

**Comment by jtopjian** [comment]
Oh, cool:

```shell
$ cat /sys/devices/virtual/dmi/id/sys_vendor
OpenStack Foundation
```

That should be more efficient than forking to `dmidecode` and removes the requirement for it.

**Comment by erichs** [comment]
Indeed! Also, there's some prior art here: https://github.com/google/cadvisor/tree/master/utils/cloudinfo

Think the cloud-detection is generally useful enough to be it's own library?

**Comment by jtopjian** [comment]
> Think the cloud-detection is generally useful enough to be it's own library?

I think if the domain of what the library did could be defined well enough, then absolutely.

Oh, one thing to beware of: The `cat /sys...` stuff is only going to work on Linux, so you'll want to put that in a `utils_linux.go` file. The `_linux.go` restricts it to only being run on Linux-binaries. [This](https://github.com/shirou/gopsutil/tree/master/docker) is the first simple example I can think of off the top of my head.

**Comment by erichs** [comment]
Ah, interesting! Thanks.

**Comment by erichs** [comment]
There's [cloudid](https://github.com/appscode/cloudid), but that essentially tries to make network connections to all of the endpoints it knows about, and fingerprint the returned results. That's interesting, and probably "good enough" if you know you're in a cloud environment, but I'd hate to have to wait on all those `net/http` calls to return just to find out they're unnecessary. There might be space for a fast, `stat`-only approach. I'll mull on that.

**Comment by jtopjian** [comment]
Agreed.

My $0.02 is that a dedicated Go library that detects the cloud platform in an efficient way would be a handy thing to have, but it certainly wouldn't block adding this feature to Elements if you wanted to do that first. The way we discussed adding it would allow a dedicated package to be swapped out fairly easily in the future.

**Comment by erichs** [comment]
Hrm, looks like this is failing with:

```
Stderr: lib/elements/elements.go:10:2: cannot find package "github.com/erichs/cloudsysfs" in any of:
	/home/travis/gopath/src/github.com/jtopjian/elements/vendor/github.com/erichs/cloudsysfs (vendor tree)
	/home/travis/.gimme/versions/go1.7.linux.amd64/src/github.com/erichs/cloudsysfs (from $GOROOT)
	/home/travis/gopath/src/github.com/erichs/cloudsysfs (from $GOPATH)
```

I'm unsure how to add this dependency to the vendored environment.

**Comment by jtopjian** [comment]
Use [govendor](https://github.com/kardianos/govendor) (as that's what has been used already for Elements). There are a bunch of vendoring tools for Go, and I believe there is an official one coming out, but I've standardized on Govendor. You'll want to do something like this:

```shell
$ cd ~/go/src/github.com/jtopjian/elements
$ govendor fetch github.com/erichs/cloudsysfs/...
$ git status
```

You'll then see it was pulled into the `vendor` directory. 

If you don't yet have cloudsysfs on github, `govendor` has an option to pull from your go path.

Let me know if it gives you any grief.

**Comment by erichs** [comment]
Worked a charm, thanks!

**Comment by erichs** [comment]
I've left it scoped to AWS only at the moment. Let me know if you'd like DO support under a separate PR, or included in this one. I don't have access to an openstack environment, but I suspect much of the existing code for AWS can be reused or refactored, depending on how closely openstack mimics the AWS api semantics.

I keep finding references to openstack supporting a `metadata.json` file and/or url. Might be worth another check to see if that can just be `json.Unmarshal`led directly.

**Comment by jtopjian** [comment]
> I keep finding references to openstack supporting a metadata.json file and/or url. Might be worth another check to see if that can just be json.Unmarshalled directly.

So it looks like `http://169.254.169.254/openstack/latest` exists. It contains much of the same info that the EC2 endpoint contains, but more openstack-oriented:

```shell
$ curl 169.254.169.254/openstack/latest   
meta_data.json                            
password                                  
vendor_data.json                          
```

There's also `config_drive`, but you must enable it when booting an instance. And then you still need to mount it within the instance. Once it's mounted, though, it acts as an on-disk metadata repository. Since the user has to mount it and the mountpoint would vary, it would probably be difficult to auto-detect.

**Comment by jtopjian** [comment]
This looks good to me. I haven't actually tested with an EC2 instance, but it works fine on a non-ec2 instance (no cloud data, but that's expected).

I'm OK with merging this if you're happy with how it looks right now. :)

File: lib/elements/aws.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	var key string
+	for _, line := range urlData {
+
+		// replace hyphens with underscores for JSON keys
+		key = strings.Replace(line, "-", "_", -1)
+
+		switch {
+		default:
+			data[key] = getData(url + line)[0]
```
Comment (jtopjian) [inline lib/elements/aws.go]: Actually, spotted something here. I did some quick modifications for OpenStack, and it crashes at this point. If I change the above to the following, everything works:

```go
    default:                   
      d := getData(url + line) 
      if len(d) > 0 {          
        data[key] = d[0]       
      }                        
```

**Comment by erichs** [comment]
Yes, I'm happy with this for now :)

Here's the behavior on AWS:
```
[ec2-user@host ~]$ ./elements_linux -v
elements version 2.5
[ec2-user@host ~]$ ./elements_linux get -p system.host
{
  "bootTime": 1490385633,
  "hostid": "3494fed0-62fc-431f-945a-02d93aa8518a",
  "hostname": "host",
  "kernelVersion": "4.4.51-40.58.amzn1.x86_64",
  "os": "linux",
  "platform": "amazon",
  "platformFamily": "rhel",
  "platformVersion": "2016.09",
  "procs": 145,
  "uptime": 972422,
  "virtualizationRole": "guest",
  "virtualizationSystem": "xen"
 }
[ec2-user@host ~]$ ./elements_linux get -p cloud.meta_data.block_device_mapping
{
  "ami": "/dev/xvda",
  "root": "/dev/xvda"
 }
```

and the behavior on DigitalOcean:
```
user@host:~$ ./elements_linux -v
elements version 2.5
user@host~$ ./elements_linux get -p system.host
{
  "bootTime": 1489779714,
  "hostid": "573E716B-7157-42F9-A03A-40694F343D9E",
  "hostname": "host",
  "kernelVersion": "4.4.0-66-generic",
  "os": "linux",
  "platform": "ubuntu",
  "platformFamily": "debian",
  "platformVersion": "14.04",
  "procs": 166,
  "uptime": 1577976,
  "virtualizationRole": "host",
  "virtualizationSystem": "kvm"
 }
user@host:~$ ./elements_linux get -p cloud
Error: Cloud provider 'digitalocean' is not supported.
```

File: lib/elements/aws.go (hunk 1 above)
Comment (erichs) [reply lib/elements/aws.go]: interesting! lemme try that.

Comment (erichs) [reply lib/elements/aws.go]: works for me!

**Comment by jtopjian** [comment]
OK, a few more thoughts, just thinking ahead for adding OpenStack and DO support:

1. Can you add `provider.OpenStack` to your `cloudsysfs/detect.go`? Very cool library, btw.
2. Could/should `crawlData` become a generic function for all known metadata urls? I was able to use it with OpenStack just fine. If so, perhaps everything in `aws.go` should be moved to `cloud.go`?
3. I'm not concerned about the OpenStack-specific URL at this time. I can look at mixing that in later. For now, the EC2-compatible URL is good to start.

**Comment by erichs** [comment]
> Can you add provider.OpenStack to your cloudsysfs/detect.go? Very cool library, btw.

Thanks! I think that's [already done, here](https://github.com/erichs/cloudsysfs/blob/master/providers/openstack.go). If it's not detecting openstack correctly, that's a bug.

> Could/should crawlData become a generic function for all known metadata urls? 

I'm not sure. It will likely work if the metadata service tries hard to ape AWS, but not so well otherwise is my guess. I thought about making it more generic, but left it as-is for now.

> I'm not concerned about the OpenStack-specific URL at this time. I can look at mixing that in later. For now, the EC2-compatible URL is good to start.

Sounds good!

**Comment by jtopjian** [comment]
> Thanks! I think that's already done, here. If it's not detecting openstack correctly, that's a bug.

I think it needs added [here](https://github.com/erichs/cloudsysfs/blob/master/detect.go) as well?

> I'm not sure. It will likely work if the metadata service tries hard to ape AWS, but not so well otherwise is my guess. I thought about making it more generic, but left it as-is for now.

Makes sense. Perhaps renaming `crawlData` and `getData` to `crawlAWSData` and `getAWSData`? I realize I'm nit-picking here, but just thinking ahead to another provider that might need to crawl as well.

**Comment by erichs** [comment]
Ah, I see, good catch! I actually renamed those functions to what you have in my DO branch, so that's probably what they should be called. I'll make those changes now.

**Comment by erichs** [comment]
TIL: `gofmt -r 'crawlData -> crawlAWSData' -w lib/elements/aws.go`. Nifty! Actually [this whole blog post](http://spf13.com/post/go-fmt/) is nifty.

**Comment by jtopjian** [comment]
That is indeed cool.

This looks great - merging!

**Comment by jtopjian** [comment]
Also, send me an email when you have a moment. My email is on my github profile.

**Comment by erichs** [comment]
Thanks!

**Pull request merged**